src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -0,0 +1,213 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+using Google.Protobuf;
+
+using OpenTelemetry.Resources;
+
+using OtlpCommon = Opentelemetry.Proto.Common.V1;
+using OtlpResource = Opentelemetry.Proto.Resource.V1;
+using OtlpTrace = Opentelemetry.Proto.Trace.V1;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation
+{
+    internal static class ActivityExtensions
+    {
+        internal static IEnumerable<OtlpTrace.ResourceSpans> ToOtlpResourceSpans(this IEnumerable<Activity> activityBatch)
+        {
+            var resourceToLibraryAndSpans = GroupByResourceAndLibrary(activityBatch);
+            var resourceSpansList = new List<OtlpTrace.ResourceSpans>(resourceToLibraryAndSpans.Count);
+
+            foreach (var resource in resourceToLibraryAndSpans)
+            {
+                var libraryList = new List<OtlpTrace.InstrumentationLibrarySpans>(resource.Value.Count);
+                foreach (var activitySourceEntry in resource.Value)
+                {
+                    var activitySource = activitySourceEntry.Key;
+                    var otlpLibrarySpans = new OtlpTrace.InstrumentationLibrarySpans
+                    {
+                        InstrumentationLibrary = new OtlpCommon.InstrumentationLibrary
+                        {
+                            Name = activitySource.Name, // Name is enforced to not be null, but it can be empty.
+                            Version = activitySource.Version ?? string.Empty, // NRE throw by proto
+                        },
+                    };
+
+                    otlpLibrarySpans.Spans.AddRange(activitySourceEntry.Value);
+                    libraryList.Add(otlpLibrarySpans);
+                }
+
+                var otlpResources = new OtlpResource.Resource();
+                otlpResources.Attributes.AddRange(
+                    resource.Key.Attributes.Select(ToOtlpAttribute));
+
+                var otlpResourceSpans = new OtlpTrace.ResourceSpans
+                {
+                    Resource = otlpResources,
+                };
+                otlpResourceSpans.InstrumentationLibrarySpans.AddRange(libraryList);
+
+                resourceSpansList.Add(otlpResourceSpans);
+            }
+
+            return resourceSpansList;
+        }
+
+        internal static OtlpTrace.Span ToOtlpSpan(this Activity activity)
+        {
+            // protobuf doesn't understand Span<T> yet: https://github.com/protocolbuffers/protobuf/issues/3431
+            Span<byte> traceIdBytes = stackalloc byte[16];
+            Span<byte> spanIdBytes = stackalloc byte[8];
+
+            activity.TraceId.CopyTo(traceIdBytes);
```

Comment:
i think its safer to check if activity.IDFormat == W3C. OT SDK sets it to W3C, but the exporter could be used independently?

Reply:
I will add the defensive code but I don't see much value in offering the exporter without the SDK.

Reply:
agree.